Remove macro impl_rare_data

kkoyung opened the pull request:
This macro does not hide any complex or unsafe implementation details, and, it's only used in two files (node.rs and element.rs). This patch removes the macro, and move the implementation in place.

Moreover, the Element::rare_data_mut function in element.rs is called by other functions, so the #[allow(dead_code)] annotation is removed. The Node::rare_data_mut function in node.rs is not called anywhere, so it is removed.

Testing: No test is needed for this cleaning up.
Fixes: #36767

kkoyung committed:
Remove macro impl_rare_data

This macro does not hide any complex or unsafe implementation details,
and, it's only used in two files (node.rs and element.rs). This patch
removes the macro, and move the implementation in place.

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

simonwuelker commented on components/script/dom/element.rs:
```diff
@@ -327,7 +327,22 @@ impl Element {
+        self.rare_data.borrow()
+    }
+
+    #[allow(dead_code)]
+    fn rare_data_mut(&self) -> RefMut<Option<Box<ElementRareData>>> {
+        self.rare_data.borrow_mut()
+    }
```
can we get rid of this `#[allow(dead_code)]` and remove the function if that causes warnings?

kkoyung replied:
It is called by other functions. I only removed the annotation, in the new commit.

simonwuelker commented on components/script/dom/node.rs:
```diff
@@ -564,7 +564,22 @@ impl Iterator for QuerySelectorIterator {
+        self.rare_data.borrow()
+    }
+
+    #[allow(dead_code)]
+    fn rare_data_mut(&self) -> RefMut<Option<Box<NodeRareData>>> {
+        self.rare_data.borrow_mut()
+    }
```
same here

kkoyung replied:
It is not called anywhere. I removed the function, in the new commit.

kkoyung committed:
Clean up the dead code

The Element::rare_data_mut function in element.rs is called by other
functions, so the #[allow(dead_code)] annotation is removed.

The Node::rare_data_mut function in node.rs is not called anywhere,
so it is removed.

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>
